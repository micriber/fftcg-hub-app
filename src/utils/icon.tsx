import React from 'react';
import {StyleSheet, Text} from 'react-native';
import reactStringReplace from 'react-string-replace';
import {compose} from './fp';
import GameIcon from '../components/icons/GameIcon';
import {ElementIconFile, GameActionIconFile} from '../enums/element';
import {makeID} from './string';

const circleEnum: {[number: string]: string} = {
  '0': '\u2775',
  '1': '\u2776',
  '１': '\u2776',
  '2': '\u2777',
  '3': '\u2778',
  '4': '\u2779',
  '5': '\u277a',
  '6': '\u277b',
  '7': '\u277c',
  '8': '\u277d',
  '9': '\u277e',
};

const styles = StyleSheet.create({
  bravoure: {fontWeight: 'bold'},
});

const _replaceTextByIcon = (
  typeToMatch: string | RegExp,
  iconReplacement: ElementIconFile | GameActionIconFile,
) => (text: string) => {
  return reactStringReplace(text, typeToMatch, () => (
    <GameIcon name={iconReplacement} key={makeID()} />
  ));
};

const _styllishBBCode = (regex: RegExp, style: Object) => (text: string) => {
  return reactStringReplace(text, regex, (match) => {
    if (match === 'br') {
      return ' ';
    } else if (match.includes('.[[br]]')) {
      return '.\n';
    } else if (match.includes('"[[br]]')) {
      return '"\n';
    } else if (match === 'Bravoure[[br]] ') {
      return (
        <Text key={makeID()} style={styles.bravoure}>
          Bravoure{'\n'}
        </Text>
      );
    } else if (match === '&middot;') {
      return '.';
    } else {
      return (
        <Text key={makeID()} style={[style]}>
          {reactStringReplace(match, '[[br]]   ', () => ' ')}
        </Text>
      );
    }
  });
};

const _styllishNumber = (regex: RegExp, style: Object) => (text: string) => {
  return reactStringReplace(text, regex, (match) => {
    const circleUnicode: string | null = circleEnum[match];
    return (
      <Text key={makeID()} style={[style]}>
        {circleUnicode}
      </Text>
    );
  });
};

const _replaceWrapper = (fn: Function) => (text: string) => fn(text);

export const replaceFireType = _replaceTextByIcon(
  /(《火》|火)/gm,
  ElementIconFile.FIRE,
);
export const replaceIceType = _replaceTextByIcon(
  /(《氷》|氷)/gm,
  ElementIconFile.ICE,
);
export const replaceWindType = _replaceTextByIcon(
  /(《風》|風)/gm,
  ElementIconFile.WIND,
);
export const replaceEarthType = _replaceTextByIcon(
  /(《土》|土)/gm,
  ElementIconFile.EARTH,
);
export const replaceLightningType = _replaceTextByIcon(
  /(《雷》|雷)/gm,
  ElementIconFile.LIGHTNING,
);
export const replaceWaterType = _replaceTextByIcon(
  /(《水》|水)/gm,
  ElementIconFile.WATER,
);
export const replaceLightType = _replaceTextByIcon(
  /(《光》|光)/gm,
  ElementIconFile.LIGHT,
);
export const replaceDarkType = _replaceTextByIcon(
  /(《闇》|闇)/gm,
  ElementIconFile.DARK,
);
export const replaceDownArrowAction = _replaceTextByIcon(
  /(《ダル》|ダル)/gm,
  GameActionIconFile.DOWN_ARROW,
);
export const replaceSpecialAction = _replaceTextByIcon(
  '《S》',
  GameActionIconFile.SPECIAL,
);

export const replaceTagS = _styllishBBCode(/\[\[[s]]]([^/]+)\[\[\/]]/gm, {
  color: '#f80',
  fontStyle: 'italic',
  textShadowColor: 'rgba(0, 0, 0, 0.2)',
  textShadowOffset: {width: -1, height: 1},
  textShadowRadius: 10,
});

export const replaceTagEX = _styllishBBCode(/\[\[ex]]([^/]+)\[\[\/]]/gm, {
  color: '#332a9d',
  fontStyle: 'italic',
  fontSize: 15,
  textShadowColor: 'rgba(0, 0, 0, 0.2)',
  textShadowOffset: {width: -1, height: 1},
  textShadowRadius: 10,
});

export const replaceTagI = _styllishBBCode(/\[\[i]]([^/]+)\[\[\/]]/gm, {
  fontStyle: 'italic',
});

export const replaceTagPBR = _styllishBBCode(
  /(\.\[\[br]] {0,3}|"\[\[br]] {0,3}|Bravoure\[\[br]] )/gm,
  {},
);
export const replaceTagBR = _styllishBBCode(/\[\[(br)]] {0,10}/gm, {});

export const replaceDot = _styllishBBCode(/(&middot;)/gm, {});

export const replaceNumber = _styllishNumber(/《([123456789１])》/gm, {});

const replaceTextByIconOrStyle = (text: string) =>
  compose(
    _replaceWrapper(replaceFireType),
    _replaceWrapper(replaceIceType),
    _replaceWrapper(replaceWindType),
    _replaceWrapper(replaceEarthType),
    _replaceWrapper(replaceLightningType),
    _replaceWrapper(replaceWaterType),
    _replaceWrapper(replaceLightType),
    _replaceWrapper(replaceDarkType),
    _replaceWrapper(replaceDownArrowAction),
    _replaceWrapper(replaceSpecialAction),
    _replaceWrapper(replaceTagEX),
    _replaceWrapper(replaceTagS),
    _replaceWrapper(replaceDot),
    _replaceWrapper(replaceTagBR),
    _replaceWrapper(replaceTagPBR),
    _replaceWrapper(replaceNumber),
    _replaceWrapper(replaceTagI),
  )(text);

export default replaceTextByIconOrStyle;
