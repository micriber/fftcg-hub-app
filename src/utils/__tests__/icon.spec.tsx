/**
 * @format
 */

import 'react-native';
import React from 'react';
import * as Icon from '../icon';
import replaceTextByIconOrStyle from '../icon';
import {render} from '@testing-library/react-native';

// More about testing in this repository:
// https://github.com/bamlab/react-native-testing/blob/master/src/pages/Home/__tests__/Home.test.tsx
describe('[Util] icon', () => {
  describe('[FIRE] type', () => {
    it('replace fire mark', async () => {
      const sample = Icon.replaceFireType('火');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace fire mark', async () => {
      const sample = Icon.replaceFireType('火 - 火');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceFireType('《1》《ダル》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[ICE] type', () => {
    it('replace ice mark', async () => {
      const sample = Icon.replaceIceType('氷');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace ice mark', async () => {
      const sample = Icon.replaceIceType('氷 - 氷');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceIceType('《1》《ダル》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[WIND] type', () => {
    it('replace wind mark', async () => {
      const sample = Icon.replaceWindType('風');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace wind mark', async () => {
      const sample = Icon.replaceWindType('風 - 風');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceWindType('《1》《ダル》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[EARTH] type', () => {
    it('replace earth mark', async () => {
      const sample = Icon.replaceEarthType('土');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace earth mark', async () => {
      const sample = Icon.replaceEarthType('土 - 土');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceEarthType('《1》《ダル》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[LIGHTNING] type', () => {
    it('replace earth mark', async () => {
      const sample = Icon.replaceLightningType('雷');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace lightning mark', async () => {
      const sample = Icon.replaceLightningType('雷 - 雷');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceLightningType('《1》《ダル》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[WATER] type', () => {
    it('replace water mark', async () => {
      const sample = Icon.replaceWaterType('水');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace water mark', async () => {
      const sample = Icon.replaceWaterType('水 - 水');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceWaterType('《1》《ダル》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[LIGHT] type', () => {
    it('replace light mark', async () => {
      const sample = Icon.replaceLightType('光');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace light mark', async () => {
      const sample = Icon.replaceLightType('光 - 光');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceLightType('《1》《ダル》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[DARK] type', () => {
    it('replace dark mark', async () => {
      const sample = Icon.replaceDarkType('闇');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace dark mark', async () => {
      const sample = Icon.replaceDarkType('闇 - 闇');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceDarkType('《1》《ダル》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[DOWN_ARROW] action', () => {
    it('replace down arrow mark', async () => {
      const sample = Icon.replaceDownArrowAction('ダル');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace down arrow mark', async () => {
      const sample = Icon.replaceDownArrowAction('ダル - ダル');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceDownArrowAction('《1》《闇》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[SPECIAL] action', () => {
    it('replace special mark', async () => {
      const sample = Icon.replaceSpecialAction('《S》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace special mark', async () => {
      const sample = Icon.replaceSpecialAction('《S》 - 《S》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceSpecialAction('《1》《闇》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[S] tag', () => {
    it('replace S mark', async () => {
      const sample = Icon.replaceTagS('[[s]] boulou [[/]]');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace S mark', async () => {
      const sample = Icon.replaceTagS('[[s]]Boulou le[[/]] - [[s]]Billy[[/]]');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceTagS('《1》《闇》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[EX] tag', () => {
    it('replace EX mark', async () => {
      const sample = Icon.replaceTagEX('[[ex]]EX BURST[[/]]');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace EX mark', async () => {
      const sample = Icon.replaceTagEX(
        '[[ex]]EX BURST 1[[/]] - [[ex]]EX BURST 2[[/]]',
      );
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceTagEX('《1》《闇》');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[BR] tag', () => {
    it('replace BR mark', async () => {
      const sample = Icon.replaceTagBR('[[ex]]EX BURST[[/]] [[br]]');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('deeply replace BR mark', async () => {
      const sample = Icon.replaceTagBR(
        '[[ex]]EX BURST 1[[/]] [[br]] - [[s]]EX BURST 2[[/]] [[br]]',
      );
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceTagBR('《1》《闇》 br');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[Number] tag', () => {
    it('replace number mark', async () => {
      const sample = Icon.replaceNumber(
        '《1》 《2》 《3》 《4》 《5》 《6》 《7》 《8》 《9》',
      );
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });

    it('replace nothing', async () => {
      const sample = Icon.replaceNumber('《闇》 br');
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });

  describe('[GLOBAL] all tag', () => {
    it('replace all mark', async () => {
      const sample = replaceTextByIconOrStyle(
        '《1》 《2》 《3》 《4》 《5》 《6》 《7》 《8》 《9》 火 氷 風 土 雷 水 光 闇 ダル 《S》 [[s]]boulou[[/]] [[ex]]EX BURST[[/]] [[br]]',
      );
      const root = render(<>{sample}</>);
      expect(root.toJSON()).toMatchSnapshot();
    });
  });
});
