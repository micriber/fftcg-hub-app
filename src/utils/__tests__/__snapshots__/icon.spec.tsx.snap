// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Util] icon [BR] tag deeply replace BR mark 1`] = `
Array [
  "[[ex]]EX BURST 1[[/]] ",
  "",
  "- [[s]]EX BURST 2[[/]] ",
  "",
  "",
]
`;

exports[`[Util] icon [BR] tag replace BR mark 1`] = `
Array [
  "[[ex]]EX BURST[[/]] ",
  "",
  "",
]
`;

exports[`[Util] icon [BR] tag replace nothing 1`] = `"《1》《闇》 br"`;

exports[`[Util] icon [DARK] type deeply replace dark mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [DARK] type replace dark mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [DARK] type replace nothing 1`] = `"《1》《ダル》"`;

exports[`[Util] icon [DOWN_ARROW] action deeply replace down arrow mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [DOWN_ARROW] action replace down arrow mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [DOWN_ARROW] action replace nothing 1`] = `"《1》《闇》"`;

exports[`[Util] icon [EARTH] type deeply replace earth mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [EARTH] type replace earth mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [EARTH] type replace nothing 1`] = `"《1》《ダル》"`;

exports[`[Util] icon [EX] tag deeply replace EX mark 1`] = `
Array [
  "",
  <Text
    style={
      Array [
        Object {
          "color": "#332a9d",
          "fontSize": 15,
          "fontStyle": "italic",
          "textShadowColor": "rgba(0, 0, 0, 0.2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": -1,
          },
          "textShadowRadius": 10,
        },
      ]
    }
  >
    EX BURST 1
  </Text>,
  " - ",
  <Text
    style={
      Array [
        Object {
          "color": "#332a9d",
          "fontSize": 15,
          "fontStyle": "italic",
          "textShadowColor": "rgba(0, 0, 0, 0.2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": -1,
          },
          "textShadowRadius": 10,
        },
      ]
    }
  >
    EX BURST 2
  </Text>,
  "",
]
`;

exports[`[Util] icon [EX] tag replace EX mark 1`] = `
Array [
  "",
  <Text
    style={
      Array [
        Object {
          "color": "#332a9d",
          "fontSize": 15,
          "fontStyle": "italic",
          "textShadowColor": "rgba(0, 0, 0, 0.2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": -1,
          },
          "textShadowRadius": 10,
        },
      ]
    }
  >
    EX BURST
  </Text>,
  "",
]
`;

exports[`[Util] icon [EX] tag replace nothing 1`] = `"《1》《闇》"`;

exports[`[Util] icon [FIRE] type deeply replace fire mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [FIRE] type replace fire mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [FIRE] type replace nothing 1`] = `"《1》《ダル》"`;

exports[`[Util] icon [GLOBAL] all tag replace all mark 1`] = `
Array [
  "",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❶
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❷
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❸
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❹
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❺
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❻
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❼
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❽
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❾
  </Text>,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " ",
  <Text
    style={
      Array [
        Object {
          "color": "#f80",
          "fontStyle": "italic",
          "textShadowColor": "rgba(0, 0, 0, 0.2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": -1,
          },
          "textShadowRadius": 10,
        },
      ]
    }
  >
    boulou
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {
          "color": "#332a9d",
          "fontSize": 15,
          "fontStyle": "italic",
          "textShadowColor": "rgba(0, 0, 0, 0.2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": -1,
          },
          "textShadowRadius": 10,
        },
      ]
    }
  >
    EX BURST
  </Text>,
  " ",
  "",
  "",
]
`;

exports[`[Util] icon [ICE] type deeply replace ice mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [ICE] type replace ice mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [ICE] type replace nothing 1`] = `"《1》《ダル》"`;

exports[`[Util] icon [LIGHT] type deeply replace light mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [LIGHT] type replace light mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [LIGHT] type replace nothing 1`] = `"《1》《ダル》"`;

exports[`[Util] icon [LIGHTNING] type deeply replace lightning mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [LIGHTNING] type replace earth mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [LIGHTNING] type replace nothing 1`] = `"《1》《ダル》"`;

exports[`[Util] icon [Number] tag replace nothing 1`] = `"《闇》 br"`;

exports[`[Util] icon [Number] tag replace number mark 1`] = `
Array [
  "",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❶
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❷
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❸
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❹
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❺
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❻
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❼
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❽
  </Text>,
  " ",
  <Text
    style={
      Array [
        Object {},
      ]
    }
  >
    ❾
  </Text>,
  "",
]
`;

exports[`[Util] icon [S] tag deeply replace S mark 1`] = `
Array [
  "",
  <Text
    style={
      Array [
        Object {
          "color": "#f80",
          "fontStyle": "italic",
          "textShadowColor": "rgba(0, 0, 0, 0.2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": -1,
          },
          "textShadowRadius": 10,
        },
      ]
    }
  >
    Boulou le
  </Text>,
  " - ",
  <Text
    style={
      Array [
        Object {
          "color": "#f80",
          "fontStyle": "italic",
          "textShadowColor": "rgba(0, 0, 0, 0.2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": -1,
          },
          "textShadowRadius": 10,
        },
      ]
    }
  >
    Billy
  </Text>,
  "",
]
`;

exports[`[Util] icon [S] tag replace S mark 1`] = `
Array [
  "",
  <Text
    style={
      Array [
        Object {
          "color": "#f80",
          "fontStyle": "italic",
          "textShadowColor": "rgba(0, 0, 0, 0.2)",
          "textShadowOffset": Object {
            "height": 1,
            "width": -1,
          },
          "textShadowRadius": 10,
        },
      ]
    }
  >
     boulou 
  </Text>,
  "",
]
`;

exports[`[Util] icon [S] tag replace nothing 1`] = `"《1》《闇》"`;

exports[`[Util] icon [SPECIAL] action deeply replace special mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [SPECIAL] action replace nothing 1`] = `"《1》《闇》"`;

exports[`[Util] icon [SPECIAL] action replace special mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [WATER] type deeply replace water mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [WATER] type replace nothing 1`] = `"《1》《ダル》"`;

exports[`[Util] icon [WATER] type replace water mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [WIND] type deeply replace wind mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  " - ",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;

exports[`[Util] icon [WIND] type replace nothing 1`] = `"《1》《ダル》"`;

exports[`[Util] icon [WIND] type replace wind mark 1`] = `
Array [
  "",
  <Image
    source="test-file-stub"
    style={
      Array [
        Object {
          "height": 20,
          "width": 20,
        },
        Object {},
      ]
    }
  />,
  "",
]
`;
